### ISSUE TITLE
[DX] Would love to see t() directly injected by WithI18n-hoc

### ISSUE THREAD

**Issue by benbender** [lingui/js-lingui#55]
Hey,

I'm playing with lingui at the moment and I'm pleased so far ;) Just two little DX-Items to reason about:

a) is there is any reason not to inject the method t() directly into the wrapped component by using WithI18n()-hoc? Would save the devs some repitive keystrokes!

b) I would love to see the naming of WithI18n changed to withI18n. Most devs are using capital first letters for classes which can be instantiated and most hocs are starting lowercase. One thing less to remember about.

Cheers & thanks

**Comment by tricoder42** [comment]
Hello,
thank you for your valuable feedback and spending some time with jsLingui.

a) From i18n object you can get a lot more stuff than just `t` method - active language, info about language (:warning: planned feature) and all formatting methods: `t`, `plural`, `select`, `selectOrdinal`, `number` and `date`. In future the lib should also support custom formatters (e.g. `range`, `relative`, etc). 

I felt that namespacing all i18n-relevant attributes and methods is better than polluting component namespace. Also, if you're using `i18n.t` too often, you can always add `const { t } = this.props.i18n` at the top of your render method.

b) That's actually pretty good idea. I saw few libs with capitalized letters, but looking and some examples (react-router, react-redux) it's clear, that lower cased first letter is more common. It also makes sense - upper case for components, lower case for functions. HOC is basically just factory for component. Would you like to add PR with `withI18n` while updating `WithI18n` to log deprecation warning?

**Comment by benbender** [comment]
Hey,

I was talking about exposing t() as a shortcut, not about replacing the i18n-object ;)
I'll try to find time for a PR today. 

Thanks for your work! 👍

**Comment by tricoder42** [comment]
Hi @benbender,
just the follow up: This problem is a bit more complex.

Babel plugin is looking for `t` (`plural`, `select`, etc.) method bounded to `i18n` object. It doesn't even check if the `i18n` object comes from `lingui` or it's your own variable. I solved similar problem with `Trans` component in React by checking the imports at the top of file.

In this case I need to figure out, how to check that `i18n` object actually comes either from `WithI18n` HOC or `react-i18n`. Problem is, that `Trans` component can access context implicitly thanks to React, while in plain JS there's no such thing as context and I have to bind translating methods to `i18n` object.

TL;DR: My suggested workaround `const { t } = this.props.i18n` won't work...

**Comment by tricoder42** [comment]
Also, **b)** is fixed in latest `lingui-react@1.1.0` release.

**Comment by tricoder42** [comment]
Closing for now, as I'm still not convinced about exposing `t` besides `i18n`.

I would rather make `const { t } = this.props.i18n` work, but that requires a bit of design change.

Thank you anyway for suggesting **b)**!
